### PR TITLE
[JENKINS-29908] Make OWASP Dependency-Check Plugin compatible con Workflow

### PR DESCRIPTION
**Pull request by recena**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

Make OWASP Dependency-Check Plugin compatible with Workflow [JENKINS-29908](https://issues.jenkins-ci.org/browse/JENKINS-29908)

```
node {
  checkout changelog: false, poll: false, scm: [$class: 'SubversionSCM', additionalCredentials: [], excludedCommitMessages: '', excludedRegions: '', excludedRevprop: '', excludedUsers: '', filterChangelog: false, ignoreDirPropChanges: false, includedRegions: '', locations: [[credentialsId: 'efaa090d-5627-4a5c-95df-f570850d89a1', depthOption: 'infinity', ignoreExternalsOption: true, local: '.', remote: 'https://server/svn/repo/trunk']], workspaceUpdater: [$class: 'UpdateUpdater']]
  sh 'mvn clean install -Dmaven.test.skip'
  step([$class: 'DependencyCheckPublisher'])
}
```

- [x] Initial source code modifications
- [x] Add an automated tests specific for Workflow compatibility

@reviewbybees

### CODE REVIEW

**Comment by recena** [comment]
@stevespringett Could you review these changes? Thanks in advance.

File: pom.xml [hunk 1]
```diff
@@ -6,7 +6,7 @@
     <parent>
         <groupId>org.jenkins-ci.plugins</groupId>
         <artifactId>plugin</artifactId>
-        <version>1.554.3</version>
+        <version>1.596.1</version>
```
Comment (jglick) [inline pom.xml]: Useful to clarify (in the PR, or in a comment in the POM itself) which core APIs you are using that requires this LTS version.

File: pom.xml [hunk 2]
```diff
@@ -49,6 +49,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
         <dependency.check.version>1.3.0</dependency.check.version>
+        <workflow-jenkins-plugin.version>1.4</workflow-jenkins-plugin.version>
```
Comment (jglick) [inline pom.xml]: You can use [up to 1.7](https://github.com/jenkinsci/workflow-plugin/blob/master/CHANGES.md#17-may-29-2015) with this LTS.

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java [hunk 3]
```diff
@@ -143,21 +154,31 @@ public String getPattern() {
...
         return new DependencyCheckProjectAction(project);
     }
 
     @Override
-    public BuildResult perform(final AbstractBuild<?, ?> build, final PluginLogger logger) throws InterruptedException, IOException {
+    public BuildResult perform(final Run<?, ?> build, final FilePath workspace, final PluginLogger logger) throws InterruptedException, IOException {
         logger.log("Collecting Dependency-Check analysis files...");
-        final FilesParser dcCollector = new FilesParser(DependencyCheckPlugin.PLUGIN_NAME, StringUtils.defaultIfEmpty(getPattern(), DEFAULT_PATTERN),
+
+        FilesParser parser = new FilesParser(DependencyCheckPlugin.PLUGIN_NAME, StringUtils.defaultIfEmpty(getPattern(), DEFAULT_PATTERN),
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java]: BTW you could make the diff a bit shorter (thus easier to review etc.) by avoiding gratuitous changes like local variable renames.

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResult.java [hunk 4]
```diff
@@ -64,15 +89,47 @@ public DependencyCheckResult(final AbstractBuild<?, ?> build, final String defau
...
+     * @deprecated see {@link #DependencyCheckResult(Run, BuildHistory, ParserResult, String, boolean)}
+     */
+    @Deprecated
     DependencyCheckResult(final AbstractBuild<?, ?> build, final BuildHistory history, final ParserResult result,
                           final String defaultEncoding, final boolean canSerialize) {
+        this((Run<?, ?>) build, history, result, defaultEncoding, canSerialize);
+
+        if (canSerialize) {
+            serializeAnnotations(result.getAnnotations());
+        }
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResult.java]: Is this not covered by the delegated-to constructor?

**Comment by jglick** [comment]
Problems with the PR builder, filed as OPS-266.

:bee:

File: pom.xml (hunk 1 above)
Comment (recena) [inline pom.xml]: @jglick This plugins requires `analysis-core` plugin and this requieres `1.596.1`

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java]: @jglick  The reason here is to follow the same naming used in other analysis plugins (Checkstyle, PMD, Android Lint, etc...). And unnecessary final modifier.

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResult.java (hunk 4 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResult.java]: @jglick Indeed. My mistake.

**Comment by jglick** [comment]
:bee:

**Comment by stevespringett** [comment]
@recena, the changes seem fine, but will need to perform testing. Waiting for jenkinsci/analysis-core-plugin#31 to be merged before I merge this pull request.

**Comment by recena** [comment]
@stevespringett Thanks for your quick reply. Sure, we should wait for https://github.com/jenkinsci/analysis-core-plugin/pull/31

I added three specific automated tests for workflow compatibility.

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java [hunk 5]
```diff
@@ -15,13 +15,11 @@
...
 
+import hudson.FilePath;
 import hudson.Launcher;
 import hudson.matrix.MatrixAggregator;
 import hudson.matrix.MatrixBuild;
-import hudson.model.AbstractBuild;
-import hudson.model.AbstractProject;
-import hudson.model.Action;
-import hudson.model.BuildListener;
+import hudson.model.*;
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckPublisher.java]: Better explicit imports.

File: src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResultAction.java [hunk 6]
```diff
@@ -39,8 +40,22 @@
      * @param owner            the associated build of this action
      * @param healthDescriptor health descriptor to use
      * @param result           the result in this build
+     *
+     * @deprecated see {@link #DependencyCheckResultAction(Run, HealthDescriptor, DependencyCheckResult)}
      */
+    @Deprecated
     public DependencyCheckResultAction(final AbstractBuild<?, ?> owner, final HealthDescriptor healthDescriptor, final DependencyCheckResult result) {
+        super((Run<?, ?>) owner, new DependencyCheckHealthDescriptor(healthDescriptor), result);
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/DependencyCheck/DependencyCheckResultAction.java]: ```this((Run<?, ?>) owner, healthDescriptor, result);``` would be better

**Comment by amuniz** [comment]
:bee:

**Comment by amuniz** [comment]
:bee:

**Comment by recena** [comment]
@reviewbybees done

**Comment by recena** [comment]
@stevespringett Hi, Static Analysis Utility [was released](https://twitter.com/jenkins_release/status/639965679341301760) yesterday. Everything is fine to merge this PR and release this plugin.

**Pull request merged**